### PR TITLE
create associated systems module

### PR DESCRIPTION
**Pull request by dbeal-eth**
* created `AssociatedSystemsModule` which is sort of like a modernized version of the addressresolver for v3. You specify an implementation address, (which can be a *router* deployed by the deployer with cannon) and then it will deploy a proxy to officially connect it to the system. Supports doing the upgrades in the same call. also supports adding unmanaged contracts (ex. when we deploy to prod, using the old SNX contract)
* change deployer ability to *not* deploy proxies. This should be done with cannon now
* change deployer to take in `modules` argument (positional), which has the ability to specify which modules to deploy. This is used by cannon to build subsets of the module directory for the 4 different systems specified above.
* add `TokenModule` and `NftModule` to `core-modules` which implements everything needed for a basic token as the title implies
* remove checks for initialization modules (though not the initialization module itself), since initialization will be a much more customized process going forward and will be done by just implementing "isInitialized" for your module.

### CODE REVIEW

File: packages/core-modules/contracts/mixins/AssociatedSystemsMixin.sol [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        return ITokenModule(_associatedSystemsStore().satellites[id].proxy);
+    }
+
+    function _getNft(bytes32 id) internal view returns (INftModule) {
+        _requireKind(id, _KIND_ERC721);
+        return INftModule(_associatedSystemsStore().satellites[id].proxy);
+    }
+
+    function _requireKind(bytes32 id, bytes32 kind) internal view {
+        bytes32 actualKind = _associatedSystemsStore().satellites[id].kind;
```
Comment (leomassazza) [inline packages/core-modules/contracts/mixins/AssociatedSystemsMixin.sol:26]: change it to emit a custom error

File: packages/core-modules/contracts/modules/AssociatedSystemsModule.sol [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        _associatedSystemsStore().satellites[id] = AssociatedSystem(endpoint, endpoint, _KIND_UNMANAGED);
+    }
+
+    function getAssociatedSystem(bytes32 id) external view override returns (address proxy, bytes32 kind) {
+        proxy = _associatedSystemsStore().satellites[id].proxy;
+        kind = _associatedSystemsStore().satellites[id].kind;
+    }
+
+    event AssociatedSystemCreated(bytes32 indexed id, address proxy, address impl);
```
Comment (leomassazza) [inline packages/core-modules/contracts/modules/AssociatedSystemsModule.sol:94]: This event is not used

File: packages/core-modules/contracts/modules/AssociatedSystemsModule.sol [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    event AssociatedSystemUpgraded(bytes32 indexed id, address proxy, address impl);
```
Comment (leomassazza) [inline packages/core-modules/contracts/modules/AssociatedSystemsModule.sol:95]: question: would it be useful to add the kind also here?

File: packages/core-modules/contracts/storage/AssociatedSystemsStorage.sol [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    struct AssociatedSystemsStore {
+        mapping(bytes32 => AssociatedSystem) satellites;
+    }
+
+    // solhint-disable-next-line func-name-mixedcase
+    function _associatedSystemsStore() internal pure returns (AssociatedSystemsStore storage store) {
+        assembly {
+            // bytes32(uint(keccak256("io.synthetix.associatedSystems")) - 1)
+            // todo
+            store.slot := 0x785a57f03c313e176889dae7a58117996eec43895df200b08edcd72475725cd5
```
Comment (leomassazza) [inline packages/core-modules/contracts/storage/AssociatedSystemsStorage.sol:20]: -1 missing: 0x785a57f03c313e176889dae7a58117996eec43895df200b08edcd72475725cd4

File: packages/core-modules/contracts/storage/NftStorage.sol [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+contract NftStorage {
+    struct NftStore {
+        bool initialized;
+        address mainProxy;
+    }
+
+    function _nftStore() internal pure returns (NftStore storage store) {
+        assembly {
+            // bytes32(uint(keccak256("io.synthetix.nft")) - 1)
+            store.slot := 0x20ae31f985b189b5ed310833a9326bfff1af705bc23a24d3e57dbbfb14628992
```
Comment (leomassazza) [inline packages/core-modules/contracts/storage/NftStorage.sol:13]: -1 missing: 0x20ae31f985b189b5ed310833a9326bfff1af705bc23a24d3e57dbbfb14628991

File: packages/core-modules/test/contracts/modules/AssociatedSystemsModule.test.js [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+        AssociatedSystemsModule.connect(user).registerUnmanagedSystem(
+          toBytes32('wohoo'),
+          owner.address
+        ),
+        'Unauthorized'
+      );
+    });
+
+    it('registers unmanaged', async () => {
+      await AssociatedSystemsModule.connect(owner).registerUnmanagedSystem(
```
Comment (leomassazza) [inline packages/core-modules/test/contracts/modules/AssociatedSystemsModule.test.js:37]: add tests to verify the events are emited correctly

**Comment by leomassazza** [comment]
`packages/deployer/task` rogue file?

File: packages/core-modules/contracts/storage/AssociatedSystemsStorage.sol (hunk 4 above)
Comment (dbeal-eth) [reply packages/core-modules/contracts/storage/AssociatedSystemsStorage.sol:20]: ooc why do we do the -1? does it have any use?

File: packages/core-modules/test/contracts/modules/AssociatedSystemsModule.test.js (hunk 6 above)
Comment (dbeal-eth) [reply packages/core-modules/test/contracts/modules/AssociatedSystemsModule.test.js]: https://github.com/Synthetixio/synthetix-v3/pull/1043/commits/3d1ab3d6afc720cf3de153c16727d2ce16356f7c

File: packages/core-modules/contracts/modules/AssociatedSystemsModule.sol (hunk 2 above)
Comment (dbeal-eth) [reply packages/core-modules/contracts/modules/AssociatedSystemsModule.sol]: https://github.com/Synthetixio/synthetix-v3/pull/1043/commits/3d1ab3d6afc720cf3de153c16727d2ce16356f7c

File: packages/core-modules/contracts/mixins/AssociatedSystemsMixin.sol (hunk 1 above)
Comment (dbeal-eth) [reply packages/core-modules/contracts/mixins/AssociatedSystemsMixin.sol:28]: https://github.com/Synthetixio/synthetix-v3/pull/1043/commits/3d1ab3d6afc720cf3de153c16727d2ce16356f7c

File: packages/core-modules/contracts/modules/AssociatedSystemsModule.sol (hunk 3 above)
Comment (dbeal-eth) [reply packages/core-modules/contracts/modules/AssociatedSystemsModule.sol]: added https://github.com/Synthetixio/synthetix-v3/pull/1043/commits/3d1ab3d6afc720cf3de153c16727d2ce16356f7c

**Comment by dbeal-eth** [comment]
im not sure why the codecov failed when it succeeded in a previous commit. In general I added more tests than existed from the previous commit

**Comment by dbeal-eth** [comment]
seems coverage is complaining that there are no tests for the testing framework changes. any way we can get those excluded?

**Comment by dbeal-eth** [comment]
ex here https://app.codecov.io/gh/Synthetixio/synthetix-v3/compare/1043/diff#D3L13

File: packages/core-modules/contracts/storage/AssociatedSystemsStorage.sol (hunk 4 above)
Comment (leomassazza) [reply packages/core-modules/contracts/storage/AssociatedSystemsStorage.sol:20]: the idea behind the -1 is that the result is less likely to be the hash of any string as suggested in EIP-1967

**Review by leomassazza — APPROVED** [review]

File: packages/deployer/utils/deployments.js [hunk 7]
```diff
@@ -136,7 +136,11 @@ function getDeploymentFolder(info) {
 }
 
 function _populateDefaults(info = {}) {
-  return { ...DeploymentInfo, ...info };
+  return {
+    folder: info.folder || DeploymentInfo.folder,
+    network: info.network || DeploymentInfo.network,
+    instance: info.instance || DeploymentInfo.instance,
+  };
```
Comment (mjlescano) [inline packages/deployer/utils/deployments.js:143]: This change is removing the possibility of renaming the default proxy contract name, if we want to keep it should be:
```suggestion
  return {
    folder: info.folder || DeploymentInfo.folder,
    network: info.network || DeploymentInfo.network,
    instance: info.instance || DeploymentInfo.instance,
    proxyContract: info.proxyContract || DeploymentInfo.proxyContract,
  };
```

Comment (mjlescano) [reply packages/deployer/utils/deployments.js:143]: Ok, disregard this comment, as we had that configuration on the defaults here but I wasn't being used, so it works correctly. Going to push a fix to completely remove it from here and just leave as a default hardhat configuration.

File: packages/core-modules/contracts/modules/NftModule.sol [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import "@synthetixio/core-contracts/contracts/utils/AddressUtil.sol";
+import "@synthetixio/core-contracts/contracts/initializable/InitializableMixin.sol";
+import "@synthetixio/core-contracts/contracts/ownership/OwnableMixin.sol";
+import "@synthetixio/core-contracts/contracts/errors/AddressError.sol";
+
+import "../storage/NftStorage.sol";
+
+import "../interfaces/INftModule.sol";
+
+contract NftModule is INftModule, ERC721, NftStorage, InitializableMixin, OwnableMixin {
```
Comment (noahlitvin) [inline packages/core-modules/contracts/modules/NftModule.sol:14]: Rename to ERC721Module?

File: packages/core-modules/contracts/modules/TokenModule.sol [hunk 9]
```diff
@@ -0,0 +1,42 @@
+//SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@synthetixio/core-contracts/contracts/initializable/InitializableMixin.sol";
+import "@synthetixio/core-contracts/contracts/ownership/OwnableMixin.sol";
+import "@synthetixio/core-contracts/contracts/token/ERC20.sol";
+import "@synthetixio/core-contracts/contracts/errors/AccessError.sol";
+import "../interfaces/ITokenModule.sol";
+
+contract TokenModule is ITokenModule, ERC20, InitializableMixin, OwnableMixin {
```
Comment (noahlitvin) [inline packages/core-modules/contracts/modules/TokenModule.sol:10]: Rename to ERC20Module?

**Review by noahlitvin — APPROVED** [review]

**Review by mjlescano — APPROVED** [review]

**Pull request merged**